File: apps/android/app/src/main/java/com/exponential/app/ui/components/WorkflowEventList.kt
package com.exponential.app.ui.components

import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.ui.platform.testTag
import androidx.compose.ui.text.font.FontFamily
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import com.exponential.app.data.db.WorkflowEventEntity
import com.exponential.app.domain.DomainContract
import com.exponential.app.domain.WireTimestamps
import com.exponential.app.ui.icons.ExpIcons
import com.exponential.app.ui.theme.GlassTokens
import com.exponential.app.ui.theme.TextEmphasis
import com.exponential.app.ui.theme.flatRow
import java.time.Instant
import java.time.ZoneId
import java.time.format.DateTimeFormatter

/**
 * EXP-1082/EXP-1068: a workflow's event log (the `workflow_events` shape).
 * Newest first (`at` desc, then id desc); each flat row = the kind's concept
 * glyph, the node's identifier as a mono lead-in ([nodeLabel]), the message on
 * one line and the local `HH:mm`. A row naming a run opens it
 * ([onOpenSession]). An empty log renders nothing.
 */
@Composable
fun WorkflowEventList(
    events: List<WorkflowEventEntity>,
    modifier: Modifier = Modifier,
    /** A node id → its identifier (`EXP-14`); null = no lead-in. */
    nodeLabel: (String) -> String? = { null },
    /** Opens the run an event names; null = rows are not clickable. */
    onOpenSession: ((String) -> Unit)? = null,
) {
    if (events.isEmpty()) return
    val ordered = events.sortedWith(
        compareByDescending<WorkflowEventEntity> { WireTimestamps.parseEpochMs(it.at) ?: 0L }
            .thenByDescending { it.id },
    )
    Column(modifier) {
        for (event in ordered) {
            val sessionId = event.sessionId
            val open = if (sessionId != null && onOpenSession != null) {
                Modifier.clickable { onOpenSession(sessionId) }
            } else {
                Modifier
            }
            Row(
                modifier = Modifier
                    .fillMaxWidth()
                    .testTag("workflow-event-row")
                    .flatRow()
                    .then(open)
                    .padding(horizontal = GlassTokens.RowPaddingH, vertical = GlassTokens.RowPaddingV),
                verticalAlignment = Alignment.CenterVertically,
            ) {
                Icon(
                    workflowEventIcon(event.kind),
                    contentDescription = null,
                    modifier = Modifier.size(14.dp),
                    tint = MaterialTheme.colorScheme.onSurface.copy(alpha = TextEmphasis.Secondary),
                )
                Spacer(Modifier.width(10.dp))
                event.nodeId?.let(nodeLabel)?.let { label ->
                    Text(
                        label,
                        style = MaterialTheme.typography.labelSmall,
                        fontFamily = FontFamily.Monospace,
                        color = MaterialTheme.colorScheme.onSurface.copy(alpha = TextEmphasis.Tertiary),
                        maxLines = 1,
                    )
                    Spacer(Modifier.width(8.dp))
                }
                Text(
                    event.message,
                    style = MaterialTheme.typography.bodySmall,
                    color = MaterialTheme.colorScheme.onSurface,
                    maxLines = 1,
                    overflow = TextOverflow.Ellipsis,
                    modifier = Modifier.weight(1f),
                )
                Spacer(Modifier.width(8.dp))
                Text(
                    workflowEventTime(event.at),
                    style = MaterialTheme.typography.labelSmall,
                    fontFamily = FontFamily.Monospace,
                    color = MaterialTheme.colorScheme.onSurface.copy(alpha = TextEmphasis.Tertiary),
                )
            }
        }
    }
}

/** The concept glyph of a contract `wfEventKind`; unknown kinds read as info. */
internal fun workflowEventIcon(kind: String): ImageVector = when (kind) {
    DomainContract.wfEventKindNodeStarted -> ExpIcons.actionRun
    DomainContract.wfEventKindRetrying -> ExpIcons.runResume
    DomainContract.wfEventKindReviewStarted,
    DomainContract.wfEventKindReviewVerdict,
    DomainContract.wfEventKindReviewNoVerdict,
    -> ExpIcons.codingInReview
    DomainContract.wfEventKindLanded,
    DomainContract.wfEventKindCompleted,
    -> ExpIcons.uiCheck
    DomainContract.wfEventKindFinalPrOpened,
    DomainContract.wfEventKindFinalPrReopened,
    -> ExpIcons.prOpen
    DomainContract.wfEventKindFailed,
    DomainContract.wfEventKindGaveUp,
    -> ExpIcons.uiWarning
    DomainContract.wfEventKindAccountPicked,
    DomainContract.wfEventKindAccountSwitched,
    DomainContract.wfEventKindWaitingReset,
    -> ExpIcons.navAccount
    DomainContract.wfEventKindQuestionAsked,
    DomainContract.wfEventKindQuestionAnswered,
    -> ExpIcons.uiHelp
    DomainContract.wfEventKindCancelled,
    DomainContract.wfEventKindSkipped,
    -> ExpIcons.uiClose
    else -> ExpIcons.uiInfo
}

private val EVENT_TIME: DateTimeFormatter = DateTimeFormatter.ofPattern("HH:mm")

/** The event's local `HH:mm`; empty for an unreadable stamp. */
private fun workflowEventTime(at: String): String {
    val ms = WireTimestamps.parseEpochMs(at) ?: return ""
    return EVENT_TIME.format(Instant.ofEpochMilli(ms).atZone(ZoneId.systemDefault()))
}

File: apps/ios/ExpUI/Sources/WorkflowEventList.swift
import ExpCore
import SwiftUI

/// EXP-1082 → EXP-1068: a workflow's EVENT LOG — the synced `workflow_events`
/// rows (the `workflow-events` shape), newest first. Each row = the kind's
/// concept glyph, the node's identifier as a mono lead-in (when `nodeLabel`
/// resolves it), the server's sentence on one line and the local `HH:mm`. A
/// row naming a run opens it through `onOpenSession`. Flat rows; an empty log
/// draws nothing.
public struct WorkflowEventList: View {
    public let events: [WorkflowEventEntity]
    /// A node id → its identifier (`EXP-14`); nil = no lead-in.
    public let nodeLabel: ((String) -> String?)?
    /// Opens the run a row names; nil = rows are not tappable.
    public let onOpenSession: ((String) -> Void)?

    public init(
        events: [WorkflowEventEntity],
        nodeLabel: ((String) -> String?)? = nil,
        onOpenSession: ((String) -> Void)? = nil
    ) {
        self.events = events
        self.nodeLabel = nodeLabel
        self.onOpenSession = onOpenSession
    }

    public var body: some View {
        VStack(spacing: 0) {
            ForEach(Self.ordered(events), id: \.id) { event in
                row(event)
            }
        }
    }

    /// Newest first: `at` descending, then id descending.
    static func ordered(_ events: [WorkflowEventEntity]) -> [WorkflowEventEntity] {
        events.sorted { a, b in
            let (ta, tb) = (instant(a.at), instant(b.at))
            return ta != tb ? ta > tb : a.id > b.id
        }
    }

    private static func instant(_ value: String) -> TimeInterval {
        WireTimestamps.parse(value)?.timeIntervalSince1970 ?? 0
    }

    /// The concept glyph a kind (contract `wfEventKind`) wears.
    static func glyph(_ kind: String) -> String {
        switch kind {
        case DomainContract.wfEventKindNodeStarted: AppIcons.actionRun
        case DomainContract.wfEventKindRetrying: AppIcons.runResume
        case DomainContract.wfEventKindReviewStarted,
             DomainContract.wfEventKindReviewVerdict,
             DomainContract.wfEventKindReviewNoVerdict: AppIcons.codingInReview
        case DomainContract.wfEventKindLanded: AppIcons.uiCheck
        case DomainContract.wfEventKindCompleted: AppIcons.uiCheck
        case DomainContract.wfEventKindFinalPrOpened,
             DomainContract.wfEventKindFinalPrReopened: AppIcons.prOpen
        case DomainContract.wfEventKindFailed,
             DomainContract.wfEventKindGaveUp: AppIcons.uiWarning
        case DomainContract.wfEventKindAccountPicked,
             DomainContract.wfEventKindAccountSwitched: AppIcons.navAccount
        case DomainContract.wfEventKindWaitingReset: AppIcons.uiClock
        case DomainContract.wfEventKindQuestionAsked,
             DomainContract.wfEventKindQuestionAnswered: AppIcons.uiHelp
        case DomainContract.wfEventKindCancelled,
             DomainContract.wfEventKindSkipped: AppIcons.uiClose
        default: AppIcons.uiInfo
        }
    }

    /// The kinds that read as a warning (amber glyph), web `WARNING_KINDS`.
    static func isWarning(_ kind: String) -> Bool {
        switch kind {
        case DomainContract.wfEventKindFailed,
             DomainContract.wfEventKindGaveUp,
             DomainContract.wfEventKindReviewNoVerdict,
             DomainContract.wfEventKindWaitingReset: true
        default: false
        }
    }

    private static let clock: DateFormatter = {
        let formatter = DateFormatter()
        formatter.locale = Locale(identifier: "en_US_POSIX")
        formatter.dateFormat = "HH:mm"
        return formatter
    }()

    /// The event's local `HH:mm`; empty when the stamp does not parse.
    static func time(_ at: String) -> String {
        WireTimestamps.parse(at).map { clock.string(from: $0) } ?? ""
    }

    @ViewBuilder
    private func row(_ event: WorkflowEventEntity) -> some View {
        if let sessionId = event.sessionId, !sessionId.isEmpty, let onOpenSession {
            Button { onOpenSession(sessionId) } label: { content(event) }
                .buttonStyle(.plain)
        } else {
            content(event)
        }
    }

    private func content(_ event: WorkflowEventEntity) -> some View {
        let label = event.nodeId.flatMap { id in nodeLabel?(id) }
        return HStack(spacing: 8) {
            AppIcon(Self.glyph(event.kind), size: 14)
                .foregroundStyle(
                    Self.isWarning(event.kind)
                        ? DesignTokens.Semantic.yellow : .white.opacity(TextOpacity.secondary)
                )
            if let label, !label.isEmpty {
                Text(label)
                    .font(.caption.monospaced())
                    .foregroundStyle(.white.opacity(TextOpacity.tertiary))
                    .lineLimit(1)
            }
            Text(event.message)
                .font(.caption)
                .foregroundStyle(.white)
                .lineLimit(1)
                .truncationMode(.tail)
            Spacer(minLength: 0)
            Text(Self.time(event.at))
                .font(.caption.monospacedDigit())
                .foregroundStyle(.white.opacity(TextOpacity.tertiary))
        }
        .padding(.horizontal, 12)
        .padding(.vertical, 10)
        .contentShape(Rectangle())
        .flatRow()
        .accessibilityIdentifier("workflow-event-\(event.kind)")
    }
}
